crates/compiler/mono/src/drop_specialization.rs

```diff
@@ -782,6 +839,121 @@ fn specialize_boxed<'a, 'i>(
     }
 }
 
+fn specialize_list<'a, 'i>(
+    arena: &'a Bump,
+    layout_interner: &'i mut STLayoutInterner<'a>,
+    ident_ids: &'i mut IdentIds,
+    environment: &mut DropSpecializationEnvironment<'a>,
+    incremented_children: &mut MutSet<Child>,
+    symbol: &Symbol,
+    item_layout: InLayout,
+    continuation: &'a Stmt<'a>,
+) -> &'a Stmt<'a> {
+    let current_length = environment.list_length.get(symbol).copied();
+
+    macro_rules! keep_original_decrement {
+        () => {{
+            let new_continuation =
+                specialize_drops_stmt(arena, layout_interner, ident_ids, environment, continuation);
+            arena.alloc(Stmt::Refcounting(ModifyRc::Dec(*symbol), new_continuation))
+        }};
+    }
+
+    match (
+        layout_interner.contains_refcounted(item_layout),
+        current_length,
+    ) {
+        (true, Some(length)) => {
+            match environment.list_children.get(symbol) {
+                Some(children) if children.len() as u64 == length => {
+                    // Only specialize lists if all children are known.
+                    // Otherwise we might have to insert an unbouned number of decrements.
+
+                    // TODO perhaps this allocation can be avoided.
+                    let children_clone = children.clone();
+
+                    // Map tracking which index of the struct is contained in which symbol.
+                    // And whether the child no longer has to be decremented.
+                    let mut index_symbols = MutMap::default();
+
+                    for index in 0..length {
+                        for (child, i) in children_clone
+                            .iter()
+                            .filter(|(_child, i)| *i == index as u64)
+                        {
+                            debug_assert!(length > *i);
+
+                            let removed = incremented_children.remove(child);
+                            index_symbols.insert(index, (*child, removed));
+
+                            if removed {
+                                break;
+                            }
+                        }
+                    }
+
+                    let new_continuation = specialize_drops_stmt(
+                        arena,
+                        layout_interner,
+                        ident_ids,
+                        environment,
+                        continuation,
+                    );
+
+                    |rc_popped: Option<
+                        fn(arena: &'a Bump, Symbol, &'a Stmt<'a>) -> &'a Stmt<'a>,
```

Comment:
I'd guess you can use a `type X = fn(...)` definition here to clean this up. it looks really weird. (contrary to some languages, type aliases in rust can be defined in nested scopes)

Reply:
also why is this construction needed (using a closure here, that is applied immediately)

Reply:
I forgot to clean it up, will do.